Check exit status and print log excerpt

orelgueta opened the pull request:
Check the exit status and print the log excerpt also when running using the `job_manager` (when running locally). A few other small improvements were included in this small PR.

Fixes #445

VictorBarbosaMartins reviewed (commented):
Thanks for updating this. I made some comments but maybe we need a second reviewer since it is the first time I am looking into this part of the code. 

I tested running corsika again through `python applications/sim_showers_for_trigger_rates.py --array 4LST --site North --primary proton --nruns 2 --nevents 10 --submit_command local` to make sure the error is printed out and it turns out it is printed out twice now. Could you please check why? (see below)

```
INFO::job_manager(l117)::_submit_local::Running script locally
ERROR::job_manager(l125)::_submit_local::Runtime error - See below the relevant part of the log file.

/workdir/external/gammasim-tools/simtools-output/sim_showers_for_trigger_rates/corsika/corsika-data/North/proton/log/log_corsika_run1_proton_North_4LST_sim_showers_for_trigger_rates.log
====================================================================

run000001: File exists

====================================================================

Traceback (most recent call last):
  File "/workdir/external/gammasim-tools/applications/sim_showers_for_trigger_rates.py", line 192, in <module>
    main()
  File "/workdir/external/gammasim-tools/applications/sim_showers_for_trigger_rates.py", line 170, in main
    shower_simulator.simulate()
  File "/workdir/external/gammasim-tools/simtools/simulator.py", line 395, in simulate
    job_manager.submit(
  File "/workdir/external/gammasim-tools/simtools/job_execution/job_manager.py", line 103, in submit
    self._submit_local(log_file)
  File "/workdir/external/gammasim-tools/simtools/job_execution/job_manager.py", line 126, in _submit_local
    raise JobExecutionError(msg)
simtools.job_execution.job_manager.JobExecutionError: Runtime error - See below the relevant part of the log file.

/workdir/external/gammasim-tools/simtools-output/sim_showers_for_trigger_rates/corsika/corsika-data/North/proton/log/log_corsika_run1_proton_North_4LST_sim_showers_for_trigger_rates.log
====================================================================

run000001: File exists

====================================================================
```

Since you already implement the `get_log_excerpt` I would suggest printing the final lines also in case of success. The way it is now, the application finishes with a not so informative message:
```
INFO::simulator(l578)::get_list_of_output_files::Getting list of output files
INFO::sim_showers_for_trigger_rates(l187)::main::List of log files exported to /workdir/external/gammasim-tools/simtools-output/sim_showers_for_trigger_rates/application-plots/log_files_proton.list
INFO::simulator(l625)::get_list_of_log_files::Getting list of log files
```
My idea would be to print also the final lines of the CORSIKA logfile, i.e., in this case:
```
 PRESENT TIME : 08.02.2023  17:00:43 UTC

 ========== END OF RUN ================================================
```

GernotMaier commented on simtools/util/general.py:
```diff
@@ -717,3 +743,27 @@ def rotate(rotation_angle, x, y):
+
+def get_log_excerpt(log_file):
+    """
+    Get an excerpt from a log file (last 30 lines).
```
Although 30 lines are good, do we want to have the 'n_print_last_lines=30' is function parameter?

orelgueta replied:
Sure, added it now.

GernotMaier reviewed (approved):
All good; one minor suggestion; nothing beyond Victor's comments.

orelgueta commented on simtools/job_execution/job_manager.py:
```diff
@@ -69,20 +73,22 @@ def test_submission_system(self):
         """
         self.run_script = str(run_script)
-        self.run_out_file = str(run_out_file).replace(".log", "")
+        self.run_out_file = str(run_out_file.parent.joinpath(run_out_file.stem))
```
True, I now convert the input parameter to be a path.

orelgueta commented on simtools/job_execution/job_manager.py:
```diff
@@ -11,6 +11,10 @@ class MissingWorkloadManager(Exception):
     """Exception for missing work load manager."""
 
 
+class JobExecutionError(Exception):
```
Sure, just did that. Though I thought we agreed at some point not to use these `__all__` lists anymore, am I wrong?

VictorBarbosaMartins replied:
Thanks. hmm I dont remember that we agreed on that. I do remember that I updated all of these lists in the big documentation PR that we had last year

orelgueta commented:
Thanks for the review! Two general comments,
- The fact it printed the excerpt twice was sort of intentional, but I removed it now. The original printout is to the logger (which one could imagine would go to a file instead of to the screen) and the other one was from the exception which would always go to the screen. However, since we don't really have this separation now and it's unclear we will ever have it, I removed the printout from the exception and point to the logger instead.
- I am not entirely sure about printing out a success message. I think we definitely do not want to print an excerpt from the log file in case of success because it's clear what you will actually get (CORSIKA/sim_telarray logs are different) and the printout might not be informative. This is in contrast to the case of failure where you would always expect the last lines of the log file to be informative. What I think could be useful is to print a success message at the application level (i.e., in this case in `sim_showers_for_trigger_rates.py`). It might be useful at the end of the application to print out a message with the time of running for example. This is separate from this PR though since it's at the application level rather than the CORSIKA level. Therefore, if you find it necessary, please open a separate issue and we can implement it (for all applications).

VictorBarbosaMartins commented on simtools/util/general.py:
```diff
@@ -432,13 +432,39 @@ def collect_final_lines(file, n_lines):
-        for line in opened_file:
-            file_in_lines.append(line)
-    collected_lines = file_in_lines[-n_lines:-1]
+    list_of_lines = []
```
In this case it seems to be a good solution, thanks.

VictorBarbosaMartins reviewed (approved):
Thanks. I agree we should open an issue for printing successful messages for the applications.
